File: esports-backend/src/main/java/com/esports/service/KycService.java
package com.esports.service;

import com.esports.dto.KycStatusResponseDTO;
import com.esports.entity.User;
import com.esports.entity.UserVerification;
import com.esports.repository.UserRepository;
import com.esports.repository.UserVerificationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Random;

@Service
public class KycService {

    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private UserVerificationRepository userVerificationRepository;
    
    @Autowired
    private S3Service s3Service;
    
    @Autowired
    private EmailService emailService;

    @Transactional
    public void sendOTP(String email) {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        String otp = String.format("%06d", new Random().nextInt(999999));
        
        emailService.sendOTP(user.getEmail(), otp);
        
        UserVerification verification = userVerificationRepository.findById(user.getId())
            .orElse(new UserVerification());
        if (verification.getUser() == null) {
            verification.setUser(user);
        }
        verification.setOtpCode(otp);
        verification.setOtpExpiry(LocalDateTime.now().plusMinutes(10));
        if (verification.getDocumentUrl() == null) {
            verification.setDocumentStatus(null);
        }
        userVerificationRepository.save(verification);
    }

    @Transactional
    public void verifyOTP(String email, String otp) {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        UserVerification verification = userVerificationRepository.findById(user.getId())
            .orElse(null);

        if (verification == null || verification.getOtpCode() == null || !verification.getOtpCode().equals(otp)) {
            throw new IllegalArgumentException("Invalid OTP");
        }

        if (verification.getOtpExpiry() == null || verification.getOtpExpiry().isBefore(LocalDateTime.now())) {
            throw new IllegalArgumentException("OTP expired");
        }

        verification.setEmailVerified(true);
        verification.setVerificationLevel(UserVerification.VerificationLevel.LEVEL_1_EMAIL);
        verification.setOtpCode(null);
        verification.setOtpExpiry(null);
        if (verification.getDocumentUrl() == null) {
            verification.setDocumentStatus(null);
        }
        userVerificationRepository.save(verification);
    }

    @Transactional
    public String uploadDocument(String email, MultipartFile file, String documentType) throws IOException {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        UserVerification verification = userVerificationRepository.findById(user.getId())
            .orElse(null);
        
        if (verification == null || verification.getVerificationLevel() != UserVerification.VerificationLevel.LEVEL_1_EMAIL) {
            throw new IllegalArgumentException("Email verification required first");
        }
        
        String folder = "kyc/user_" + user.getId();
        Map<String, String> s3Data = s3Service.upload(file, folder);
        String documentUrl = s3Data.get("url");
        
        verification.setDocumentUrl(documentUrl);
        verification.setDocumentType(documentType);
        verification.setDocumentUploadedAt(LocalDateTime.now());
        verification.setDocumentStatus(UserVerification.DocumentStatus.PENDING);
        
        userVerificationRepository.save(verification);
        
        return documentUrl;
    }

    public KycStatusResponseDTO getKYCStatus(String email) {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        UserVerification verification = userVerificationRepository.findById(user.getId())
            .orElse(null);
        
        if (verification != null && verification.getDocumentUrl() == null && verification.getDocumentStatus() != null) {
            verification.setDocumentStatus(null);
            userVerificationRepository.save(verification);
        }
        
        return new KycStatusResponseDTO(
            verification != null ? verification.getVerificationLevel().toString() : "UNVERIFIED",
            verification != null ? verification.isEmailVerified() : false,
            verification != null && verification.getDocumentUrl() != null,
            verification != null && verification.getDocumentType() != null ? verification.getDocumentType() : "",
            verification != null && verification.getDocumentUrl() != null ? 
                (verification.getDocumentStatus() != null ? verification.getDocumentStatus().toString() : "PENDING") : "NOT_UPLOADED"
        );
    }
}

File: esports-backend-dotnet/Upkart1/Services/KycService.cs
using Microsoft.EntityFrameworkCore;
using Upkart1.Data;
using Upkart1.DTO;
using Upkart1.Entities;
using Upkart1.Entities.Enums;

namespace Upkart1.Services
{
    public class KycService : IKycService
    {
        private readonly ApplicationDbContext _context;
        private readonly IS3Service _s3Service;
        private readonly IEmailService _emailService;

        public KycService(ApplicationDbContext context, IS3Service s3Service, IEmailService emailService)
        {
            _context = context;
            _s3Service = s3Service;
            _emailService = emailService;
        }

        public void SendOTP(string email)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var otp = new Random().Next(100000, 999999).ToString("D6");

            _emailService.SendOTP(user.Email, otp);

            var verification = _context.UserVerifications.FirstOrDefault(v => v.UserId == user.Id)
                ?? new UserVerification { UserId = user.Id };

            if (verification.UserId == 0)
                _context.UserVerifications.Add(verification);

            verification.OtpCode = otp;
            verification.OtpExpiry = DateTime.Now.AddMinutes(10);
            if (verification.DocumentUrl == null)
                verification.DocumentStatus = null;

            _context.SaveChanges();
        }

        public void VerifyOTP(string email, string otp)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var verification = _context.UserVerifications.FirstOrDefault(v => v.UserId == user.Id);

            if (verification == null || verification.OtpCode == null || verification.OtpCode != otp)
                throw new ArgumentException("Invalid OTP");

            if (verification.OtpExpiry == null || verification.OtpExpiry < DateTime.Now)
                throw new ArgumentException("OTP expired");

            verification.EmailVerified = true;
            verification.VerificationLevel = VerificationLevel.Level1Email;
            verification.OtpCode = null;
            verification.OtpExpiry = null;
            if (verification.DocumentUrl == null)
                verification.DocumentStatus = null;

            _context.SaveChanges();
        }

        public async Task<string> UploadDocument(string email, IFormFile file, string documentType)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var verification = _context.UserVerifications.FirstOrDefault(v => v.UserId == user.Id);

            if (verification == null || verification.VerificationLevel != VerificationLevel.Level1Email)
                throw new ArgumentException("Email verification required first");

            var folder = $"kyc/user_{user.Id}";
            var s3Data = await _s3Service.Upload(file, folder);
            var documentUrl = s3Data["url"];

            verification.DocumentUrl = documentUrl;
            verification.DocumentType = documentType;
            verification.DocumentUploadedAt = DateTime.Now;
            verification.DocumentStatus = DocumentStatus.Pending;

            _context.SaveChanges();

            return documentUrl;
        }

        public KycStatusResponseDTO GetKYCStatus(string email)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var verification = _context.UserVerifications.FirstOrDefault(v => v.UserId == user.Id);

            if (verification != null && verification.DocumentUrl == null && verification.DocumentStatus != null)
            {
                verification.DocumentStatus = null;
                _context.SaveChanges();
            }

            return new KycStatusResponseDTO(
                verification?.VerificationLevel.ToString() ?? "UNVERIFIED",
                verification?.EmailVerified ?? false,
                verification?.DocumentUrl != null,
                verification?.DocumentType ?? "",
                verification?.DocumentUrl != null ?
                    (verification.DocumentStatus?.ToString() ?? "PENDING") : "NOT_UPLOADED"
            );
        }
    }
}
